src/box/sql/fk_constraint.c

```diff
@@ -1,633 +0,0 @@
-	 * parent table.
-	 */
-	if (incr_count < 0) {
-		sqlVdbeAddOp2(v, OP_FkIfZero, false, ok_label);
-	}
-	struct field_link *link = fk_def->links;
-	for (uint32_t i = 0; i < fk_def->field_count; ++i, ++link) {
-		int reg = link->child_field + reg_data + 1;
-		sqlVdbeAddOp2(v, OP_IsNull, reg, ok_label);
-	}
-	uint32_t field_count = fk_def->field_count;
-	/*
-	 * If the parent table is the same as the child table, and
-	 * we are about to increment the constraint-counter (i.e.
-	 * this is an INSERT operation), then check if the row
-	 * being inserted matches itself. If so, do not increment
-	 * the constraint-counter.
-	 *
-	 * If any of the parent-key values are NULL, then the row
-	 * cannot match itself. So set JUMPIFNULL to make sure we
-	 * do the OP_Found if any of the parent-key values are
-	 * NULL (at this point it is known that none of the child
-	 * key values are).
-	 */
-	if (fk_constraint_is_self_referenced(fk_def) && incr_count == 1) {
-		int jump = sqlVdbeCurrentAddr(v) + field_count + 1;
-		link = fk_def->links;
-		for (uint32_t i = 0; i < field_count; ++i, ++link) {
-			int chcol = link->child_field + 1 + reg_data;
-			int pcol = link->parent_field + 1 + reg_data;
-			sqlVdbeAddOp3(v, OP_Ne, chcol, jump, pcol);
-			sqlVdbeChangeP5(v, SQL_JUMPIFNULL);
-		}
-		sqlVdbeGoto(v, ok_label);
-	}
-	/**
-	 * Inspect a parent table with OP_Found.
-	 * We mustn't make it for a self-referenced table since
-	 * it's tuple will be modified by the update operation.
-	 * And since the foreign key has already detected a
-	 * conflict, fk counter must be increased.
-	 */
-	if (!(fk_constraint_is_self_referenced(fk_def) && is_update)) {
-		int temp_regs = sqlGetTempRange(parse_context, field_count);
-		int rec_reg = sqlGetTempReg(parse_context);
-		vdbe_emit_open_cursor(parse_context, cursor, referenced_idx,
-				      parent);
-		link = fk_def->links;
-		for (uint32_t i = 0; i < field_count; ++i, ++link) {
-			sqlVdbeAddOp2(v, OP_Copy,
-					  link->child_field + 1 + reg_data,
-					  temp_regs + i);
-		}
-		sqlVdbeAddOp3(v, OP_MakeRecord, temp_regs, field_count,
-			      rec_reg);
-		sqlVdbeAddOp4Int(v, OP_Found, cursor, ok_label, rec_reg, 0);
-		sqlReleaseTempReg(parse_context, rec_reg);
-		sqlReleaseTempRange(parse_context, temp_regs, field_count);
-	}
-	if (parse_context->pToplevel == NULL && !parse_context->isMultiWrite) {
-		/*
-		 * If this is an INSERT statement that will insert
-		 * exactly one row into the table, raise a
-		 * constraint immediately instead of incrementing
-		 * a counter. This is necessary as the VM code is
-		 * being generated for will not open a statement
-		 * transaction.
-		 */
-		assert(incr_count == 1);
-		const char *err = tt_sprintf(tnt_errcode_desc(ER_SQL_EXECUTE),
-					     "FOREIGN KEY constraint failed");
-		sqlVdbeAddOp4(v, OP_SetDiag, ER_SQL_EXECUTE, 0, 0, err,
-			      P4_STATIC);
-		sqlVdbeAddOp1(v, OP_Halt, -1);
-	} else {
-		sqlVdbeAddOp2(v, OP_FkCounter, false, incr_count);
```

Comment:
`OP_FkCounter` can be deleted.

Reply:
Thanks! Dropped `OP_FkCounter` and `OP_FkIfZero`.